File: mirea/cources/2026-2027/aii-july/03/contest/C-2.py
# Данное решение работает за линейное времян - за O(n+q). Мы храним одновременно:
# 1) стек из элементов;
# 2) стек на минимум;
# 3) стек на максимум;
# 4) хеш-таблицу с количеством каждого элемента.
# Из хеш-таблицы удаляем значения, когда их количество достигло нуля. Таким образом,
# размер хеш-таблицы и будет количеством различных элементов. А минимум и максимум
# берём с верхушки стеков.
inf = 1000000007


def solve():
    # читаем данные
    n = int(input())
    cnt = {}
    stVal, stMin, stMax = [], [], []
    # добавим бесконечности:
    stVal.append(inf)
    stMin.append(+inf)
    stMax.append(-inf)
    # теперь добавим нормальные элементы:
    i = 0
    while i < n:
        for x in map(int, input().split()):
            stVal.append(x)
            stMin.append(min(stMin[-1], x))
            stMax.append(max(stMax[-1], x))
            cnt[x] = cnt.get(x, 0) + 1
            i += 1
            if i == n:
                break
    # обрабатываем запросы
    q = int(input())
    while q > 0:
        query = input().split()
        c = query[0]
        if c == '+':
            x = int(query[1])
            stVal.append(x)
            stMin.append(min(stMin[-1], x))
            stMax.append(max(stMax[-1], x))
            cnt[x] = cnt.get(x, 0) + 1
        elif c == '?':
            print(stMin[-1], stMax[-1], len(cnt))
        elif c == '-':
            # сначала удалим из стека:
            last = stVal[-1]
            stVal.pop()
            stMin.pop()
            stMax.pop()
            # теперь аккуратно удалим из хеш-таблицы:
            cnt[last] -= 1  # уменьшаем количество
            if cnt[last] == 0:  # стало равным нулю => удаляем
                del cnt[last]
        q -= 1


tt = int(input())
for _ in range(tt):
    solve()

File: mirea/cources/2026-2027/aii-july/03/contest/C-2.cpp
// Данное решение работает за линейное времян - за O(n+q). Мы храним одновременно:
// 1) стек из элементов;
// 2) стек на минимум;
// 3) стек на максимум;
// 4) хеш-таблицу с количеством каждого элемента.
// Из хеш-таблицы удаляем значения, когда их количество достигло нуля. Таким образом,
// размер хеш-таблицы и будет количеством различных элементов. А минимум и максимум
// берём с верхушки стеков.
#include <bits/stdc++.h>
using namespace std;
const int inf = 1000000007;
void solve() {
    // читаем данные
    int n; cin >> n;
    unordered_map<int, int> cnt;
    stack<int> stVal, stMin, stMax;
    // добавим бесконечности:
    stVal.push(inf);
    stMin.push(+inf);
    stMax.push(-inf);
    // теперь добавим нормальные элементы:
    for (int i = 0; i < n; i++) {
        int x; cin >> x;
        stVal.push(x);
        stMin.push(min(stMin.top(), x));
        stMax.push(max(stMax.top(), x));
        cnt[x]++;
    }
    // обрабатываем запросы
    int q; cin >> q;
    while (q --> 0) {
        char c; cin >> c;
        if (c == '+') {
            int x; cin >> x;
            stVal.push(x);
            stMin.push(min(stMin.top(), x));
            stMax.push(max(stMax.top(), x));
            cnt[x]++;
        } else if (c == '?') {
            cout << stMin.top() << ' ' << stMax.top() << ' ' << cnt.size() << '\n';
        } else if (c == '-') {
            // сначала удалим из стека:
            int last = stVal.top();
            stVal.pop(); stMin.pop(); stMax.pop();
            // теперь аккуратно удалим из хеш-таблицы:
            auto iter = cnt.find(last);
            iter->second--; // уменьшаем количество
            if (iter->second == 0) // стало равным нулю => удаляем
                cnt.erase(iter);
        }
    }
}
main() {
    int tt; cin >> tt;
    while (tt --> 0) solve();
}
